components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![deny(missing_docs)]
+
+//! A memory cache Implements the logic specified in http://tools.ietf.org/html/rfc7234
+//! and http://tools.ietf.org/html/rfc7232.
+
+use fetch::methods::DoneChannel;
+use hyper::header;
+use hyper::header::ContentType;
+use hyper::header::Headers;
+use hyper::method::Method;
+use hyper::status::StatusCode;
+use hyper_serde::Serde;
+use net_traits::{Metadata, FetchMetadata, FilteredMetadata};
+use net_traits::request::Request;
+use net_traits::response::{Response, ResponseBody};
+use servo_url::ServoUrl;
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::{Arc, Mutex};
+use time;
+use time::{Duration, Tm};
+use url::Url;
+
+
+/// The key used to differentiate requests in the cache.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct CacheKey {
+    url: ServoUrl
```

Comment:
I saw that you originally included headers in the key, then later removed them. What was the issue, exactly? Could we just omit the vary-related headers when comparing them and checking the key?

Reply:
One problem was that I ended up doing a lot of looping through all entries and comparing `key.url()` with `request.url()`, for example in the context of vary, since the headers will change from request to request, while the url stays the same, and we still need to match those.

The problem with excluding vary headers is that on the request side you don't know which headers it will be(it is only the response that contains a `Vary` header, containing the name of the relevant request headers). Also, if the second requests contains additional headers, unrelated to the Vary check, you need to ignore those. If the headers are in the key, such request will be excluded, hence I was looping and comparing url.

Same issue with invalidation, one needs to invalidate based on the url(or value of `Location` or `Content-Location` header, which are urls as well).

I did also run into issues related to request `cache-control` headers, which I did initially successfully exclude when building they key. 

So based on those use cases, it felt more natural to have urls as keys. The spec also points in that direction.